processing/src/main/java/io/druid/query/groupby/epinephelinae/ParallelCombiner.java

```diff
@@ -0,0 +1,496 @@
+    final List<CloseableIterator<Entry<KeyType>>> childIteratorsOfNextLevel = new ArrayList<>();
+    final List<Future> combineFutures = new ArrayList<>();
+
+    // The below algorithm creates the combining nodes of the current level. It first checks that the number of children
+    // to be combined together is 1. If it is, the intermediate combining node for that child is not needed. Instead, it
+    // can be directly connected to a node of the parent level. Here is an example of generated tree when
+    // numLeafNodes = 6 and combineDegree = 2.
+    //      o
+    //     / \
+    //    o   \
+    //   / \   \
+    //  o   o   o
+    // / \ / \ / \
+    // o o o o o o
+    //
+    // We can expect that the aggregates can be combined as early as possible because the tree is built in a bottom-up
+    // manner.
+
+    for (int i = 0; i < numChildLevelIterators; i += combineDegree) {
+      if (i < numChildLevelIterators - 1) {
+        final List<? extends CloseableIterator<Entry<KeyType>>> subIterators = childIterators.subList(
+            i,
+            Math.min(i + combineDegree, numChildLevelIterators)
+        );
+        final Pair<CloseableIterator<Entry<KeyType>>, Future> iteratorAndFuture = runCombiner(
+            subIterators,
+            bufferSupplier.get(),
+            combiningFactories,
+            dictionary
+        );
+
+        childIteratorsOfNextLevel.add(iteratorAndFuture.lhs);
+        combineFutures.add(iteratorAndFuture.rhs);
+      } else {
+        // If there remains one child, it can be directly connected to a node of the parent level.
+        childIteratorsOfNextLevel.add(childIterators.get(i));
+      }
+    }
+
+    if (childIteratorsOfNextLevel.size() == 1) {
+      // This is the root
+      return Pair.of(childIteratorsOfNextLevel, combineFutures);
+    } else {
+      // Build the parent level iterators
+      final Pair<List<CloseableIterator<Entry<KeyType>>>, List<Future>> parentIteratorsAndFutures =
+          buildCombineTree(childIteratorsOfNextLevel, bufferSupplier, combiningFactories, combineDegree, dictionary);
+      combineFutures.addAll(parentIteratorsAndFutures.rhs);
+      return Pair.of(parentIteratorsAndFutures.lhs, combineFutures);
+    }
+  }
+
+  private Pair<CloseableIterator<Entry<KeyType>>, Future> runCombiner(
+      List<? extends CloseableIterator<Entry<KeyType>>> iterators,
+      ByteBuffer combineBuffer,
+      AggregatorFactory[] combiningFactories,
+      List<String> dictionary
+  )
+  {
+    final SettableColumnSelectorFactory settableColumnSelectorFactory =
+        new SettableColumnSelectorFactory(combiningFactories);
+    final StreamingMergeSortedGrouper<KeyType> grouper = new StreamingMergeSortedGrouper<>(
+        Suppliers.ofInstance(combineBuffer),
+        combineKeySerdeFactory.factorizeWithDictionary(dictionary),
+        settableColumnSelectorFactory,
+        combiningFactories
+    );
+    grouper.init();
```

Comment:
Please leave a comment why it should be outside lambda

Reply:
Sure, added a comment.